lib/api/api_auth.go

```diff
@@ -82,33 +82,40 @@ func basicAuthAndSessionMiddleware(cookieName string, guiCfg config.GUIConfigura
 			}
 		}
 
-		l.Debugln("Sessionless HTTP request with authentication; this is expensive.")
-
-		username, password, ok := r.BasicAuth()
-		if !ok {
-			unauthorized(w)
+		if username, ok := attemptBasicAuth(r, guiCfg, ldapCfg, evLogger); ok {
+			createSession(cookieName, username, guiCfg, evLogger, w, r)
+			next.ServeHTTP(w, r)
 			return
 		}
 
-		authOk := auth(username, password, guiCfg, ldapCfg)
-		if !authOk {
-			usernameIso := string(iso88591ToUTF8([]byte(username)))
-			passwordIso := string(iso88591ToUTF8([]byte(password)))
-			authOk = auth(usernameIso, passwordIso, guiCfg, ldapCfg)
-			if authOk {
-				username = usernameIso
-			}
-		}
+		unauthorized(w)
+	})
+}
 
-		if !authOk {
-			emitLoginAttempt(false, username, r.RemoteAddr, evLogger)
-			unauthorized(w)
-			return
+func attemptBasicAuth(r *http.Request, guiCfg config.GUIConfiguration, ldapCfg config.LDAPConfiguration, evLogger events.Logger) (string, bool) {
+	username, password, ok := r.BasicAuth()
+	if !ok {
+		return "", false
+	}
+
+	l.Debugln("Sessionless HTTP request with authentication; this is expensive.")
+
+	authOk := auth(username, password, guiCfg, ldapCfg)
+	if !authOk {
+		usernameIso := string(iso88591ToUTF8([]byte(username)))
+		passwordIso := string(iso88591ToUTF8([]byte(password)))
+		authOk = auth(usernameIso, passwordIso, guiCfg, ldapCfg)
+		if authOk {
+			username = usernameIso
 		}
+	}
```

Comment:
Contrary to the above, here I'd check for `authOk` and then immediately return. While it doesn't check for failure, it keeps things flatter. If the check stays at `!authOk`, I advocate for making it `!authOk` on the inner condition as well and return from there.

Reply:
That wouldn't work if `authOk` was already true, though. But this highlights that it's not obvious what this section of code does, so I'll try refactoring it with some kind of `authWithLatin1Fallback` function to make the flow easier to follow.

Reply:
Coming back to this, I don't know what I meant wouldn't work. :laughing: I think I ended up doing what you suggested, see 4b7c3f268b19e430ec8cbcccbea6b3f0537b5365.